activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -99,21 +101,23 @@ def select_all(arel, name = nil, binds = [], preparable: nil)
       private
 
         def cache_sql(sql, name, binds)
-          result =
-            if @query_cache[sql].key?(binds)
-              ActiveSupport::Notifications.instrument(
-                "sql.active_record",
-                sql: sql,
-                binds: binds,
-                name: name,
-                connection_id: object_id,
-                cached: true,
-              )
-              @query_cache[sql][binds]
-            else
-              @query_cache[sql][binds] = yield
-            end
-          result.dup
+          @lock.synchronize do
```

Comment:
> This code only runs in the fixture setup and teardown so it does not
affect real production databases.

@eileencodes @matthewd Is this lock and the lock around the `connection_id` necessary outside of the test environment?

Reply:
No.. my theory was that acquiring an uncontended lock wouldn't be noticeably slower than checking whether the lock was needed (given that we're about to perform IO anyway). I guess the fact you're asking suggests I was wrong?

Reply:
Yea I started noticing it in our flamegraphs but the overhead doesn't contribute significantly when I tried to benchmark it.